### DISCUSSION TITLE
Why is there a subtraction of 2 in the calculation for the maximum memory?

### DISCUSSION THREAD

**Discussion by gugaio** [bitsandbytes-foundation/bitsandbytes#1023] [Q&A]
Hi guys, i've done extensive research, but unfortunately, I can't find any explanation why is there a subtraction of 2 in max memory calculation mentioned in the README and example code.

`max_memory={
    i: f'{int(torch.cuda.mem_get_info(i)[0]/1024**3)-2}GB'
    for i in range(torch.cuda.device_count())
  }`

Could someone kindly provide an explanation for the purpose behind this subtraction?"

**Comment by gyanu2507** [comment]
The `-2` is a fudge factor in gigabytes, not part of the quantizer. `torch.cuda.mem_get_info()[0]` is free VRAM *right now*. `device_map="auto"` / accelerate then tries to pack weights up to `max_memory`. CUDA context, residual activations, KV cache, and the bitsandbytes workspace are not in that "free" number the way you wish they were.

Leaving ~2 GB unused per GPU is how the README stops the load from succeeding and then OOMing on the first forward. On a 24 GB card you would write `22GB`, not `24GB`.

Tune it. 2 GB is conservative on 80 GB GPUs (1 GB is often enough). On an 8 GB card, `-2` is a huge slice and you may need `-1` or a manual `max_memory={0: "6GiB"}`. The subtraction is not a bitsandbytes kernel constant.
